# Cost of Plutocracy

The aim of this article is to analyze some of the biggest DAOs in the crypto space and see how many of them are influenced by token holders which own a disproportionate amount of tokens compared to most voters.

In [1]:
# sets up the pynb environment
import os
import sys

import pandas as pd
import numpy as np


module_path = os.path.abspath(os.path.join(".."))
if module_path not in sys.path:
    sys.path.append(module_path)

from libs.data_processing.filters import get_quartile_by_vp
from libs.data_processing.statistics import (
    get_number_of_whales_to_all_voters_ratio,
    get_score_comparisons,
)


The data used for this article came from data sources that specialize in providing data on DAO governance, such as DeepDAO and Snapshot. Coingecko was also used to find market data for DAO native tokens.

Using this data, two spreadsheets were compiled which act as the local database for this analysis. Each spreadsheet contains each voter's choice and voting power for the last five proposals each DAO opened. **One spreadsheet filters out "whales"** which, in the context of this analysis, are voters of each proposal which have voting power **at or above the 95th percentile of voting power for that proposal**.

In [2]:
plutocracy_report_data = pd.read_excel(
    "../plutocracy_report.xls", sheet_name=None, engine="openpyxl"
)
plutocracy_report_data_filtered = pd.read_excel(
    "../plutocracy_report_filtered.xls", sheet_name=None, engine="openpyxl"
)


With this, we can begin asking ourselves some questions. Such as: **How many whales have voted in the last five proposals for each of the DAOs we analyzed?**

In [3]:
voting_ratios = get_number_of_whales_to_all_voters_ratio(
    plutocracy_report_data, plutocracy_report_data_filtered
)


In [4]:
pd.DataFrame(
    [list(result.items())[0][1] for result in voting_ratios],
    index=[list(result.items())[0][0] for result in voting_ratios],
    columns=[
        "# of whales",
        "all voters",
        "avg vp for non-whales",
        "avg vp for whales",
        "avg cost of vote",
    ],
)


# of whales  all voters  avg vp for non-whales  \
Uniswap                565       12120               1.284866   
ENS                     74        1609               3.521251   
Lido                   129        2396               2.996389   
Frax                     1          27          245572.420746   
Decentraland            16         264              94.580925   
Curve Finance            8         181            6146.094560   
Radicle                  1          28          213089.903084   
Euler                    3          82            1050.633823   
Hop                     48         944             267.985913   

               avg vp for whales  avg cost of vote  
Uniswap             6.275766e+04          6.095784  
ENS                 3.356937e+04         14.895998  
Lido                7.571993e+05          1.455704  
Frax                3.806272e+06         10.240349  
Decentraland        7.613737e+04          0.749201  
Curve Finance       2.509741e+06          2.485927  
Radicle             2.830379e+06          4.079941  
Euler               6.614156e+04          5.136914  
Hop                 3.610917e+05          0.083825

Here we can clearly see that for the DAOs which we choose to analyze the top 5% of voters have a clear, disproportionate amount of voting power comparared to the average voter.

We can also gain some insight at the economic might that these whales hold for each DAO at the time of voting by taking a look at the average cost of each vote for each DAO's native token at the time these proposals were active.

---

We move on to taking at what DAO participation would look like without these whale's involvement. We do so by comparing the scores of each proposal and checking whether or not the outcome of the proposal was changed after filtering out whales.

We check whether or not a propsal's outcome has changed simply by checking if the largest vote choice score has changed after filtering out whales, more specifically, in python we do:
```python
has_changed_outcome = not unfiltered_winning_choice_index == filtered_winning_choice_index
```

In [5]:
score_diffrences = get_score_comparisons(
    plutocracy_report_data, plutocracy_report_data_filtered
)


In [6]:
score_diffrences_dfs = dict()

for score_diffrence in score_diffrences:
    for organization, data in score_diffrence.items():
        data: dict[str, list] = data
        items = data.items()
        score_diffrences_dfs[organization] = pd.DataFrame(
            [score_data for _, score_data in items],
            index=pd.Index(
                ([proposal_id for proposal_id, _ in items]), name="Proposal ID"
            ),
            columns=[
                "type",
                "choices",
                "score_diffrences",
                "total_vp",
                "has_changed_outcome",
                "changed_outcome_winner"
            ],
        )
        score_diffrences_dfs[organization]["total_vp"] = score_diffrences_dfs[
            organization
        ]["total_vp"].apply("{:.9f}".format)


First, let's look at Uniswap

In [7]:
score_diffrences_dfs["Uniswap"]

type  \
Proposal ID                                                         
0x6b8df360fdf73085b21fdf5eef9f85916fbde95621a3d...  single-choice   
0xa3c7247632d3a08e4e73a63867908d0cd4402f48fa673...  single-choice   
0xfd3d3807bd2a6eda1327c311b83de235061d39ff1bdfb...  single-choice   
0x387b5e811e9673359b98729af5c7f14a50b7d0ea9b4dc...  single-choice   

                                                                                              choices  \
Proposal ID                                                                                             
0x6b8df360fdf73085b21fdf5eef9f85916fbde95621a3d...     ['Wormhole', 'LayerZero', 'Celer', 'deBridge']   
0xa3c7247632d3a08e4e73a63867908d0cd4402f48fa673...                                      ['Yes', 'No']   
0xfd3d3807bd2a6eda1327c311b83de235061d39ff1bdfb...  ['add Overnight Finance', 'add DeFiEdge', 'add...   
0x387b5e811e9673359b98729af5c7f14a50b7d0ea9b4dc...          ['Yes - Enact changes', 'No - No change']   

                                                                                     score_diffrences  \
Proposal ID                                                                                             
0x6b8df360fdf73085b21fdf5eef9f85916fbde95621a3d...  [9546794.425500717, 8863809.305899736, 540.859...   
0xa3c7247632d3a08e4e73a63867908d0cd4402f48fa673...            [20056524.265649345, 4927056.974342918]   
0xfd3d3807bd2a6eda1327c311b83de235061d39ff1bdfb...  [10054.585075893452, 2708025.0728700403, 3412....   
0x387b5e811e9673359b98729af5c7f14a50b7d0ea9b4dc...             [30508464.50667036, 34.17723847808492]   

                                                              total_vp  \
Proposal ID                                                              
0x6b8df360fdf73085b21fdf5eef9f85916fbde95621a3d...  23548080.599735759   
0xa3c7247632d3a08e4e73a63867908d0cd4402f48fa673...  24994560.158377878   
0xfd3d3807bd2a6eda1327c311b83de235061d39ff1bdfb...   2728492.255083876   
0x387b5e811e9673359b98729af5c7f14a50b7d0ea9b4dc...  61343398.710826278   

                                                    has_changed_outcome  \
Proposal ID                                                               
0x6b8df360fdf73085b21fdf5eef9f85916fbde95621a3d...                False   
0xa3c7247632d3a08e4e73a63867908d0cd4402f48fa673...                False   
0xfd3d3807bd2a6eda1327c311b83de235061d39ff1bdfb...                 True   
0x387b5e811e9673359b98729af5c7f14a50b7d0ea9b4dc...                False   

                                                   changed_outcome_winner  
Proposal ID                                                                
0x6b8df360fdf73085b21fdf5eef9f85916fbde95621a3d...               Wormhole  
0xa3c7247632d3a08e4e73a63867908d0cd4402f48fa673...                    Yes  
0xfd3d3807bd2a6eda1327c311b83de235061d39ff1bdfb...  add Overnight Finance  
0x387b5e811e9673359b98729af5c7f14a50b7d0ea9b4dc...    Yes - Enact changes

We observe that out of the last five proposals only one had their outcome changed after removing whales: The proposal to [add a Liquidity Mining Manager for the Optimism-Uniswap Liquidty Mining Program](https://snapshot.org/#/uniswap/proposal/0xfd3d3807bd2a6eda1327c311b83de235061d39ff1bdfb616c9f9b0d367c3ac2c).

In this case, "Overnight Finance" would've been the chosen LMM if it weren't for whale interevention, instead "DeFiEdge" was chosen for this campaign. From the data, we can see that without whale intervention, DeFiEdge would've had ~2,708,025 less votes out of the total ~2,728,492 voting power for the entire proposal. That is ~99.3% of voting power allocated to choosing "DeFiEdge".

In [8]:
propsal_choices = plutocracy_report_data['Uniswap'][plutocracy_report_data['Uniswap']['proposal_id'] == '0xfd3d3807bd2a6eda1327c311b83de235061d39ff1bdfb616c9f9b0d367c3ac2c'].iloc[0]['proposal_choices']
proposal_score_diffrences = score_diffrences_dfs["Uniswap"].loc["0xfd3d3807bd2a6eda1327c311b83de235061d39ff1bdfb616c9f9b0d367c3ac2c"]["score_diffrences"]
proposal_scores = plutocracy_report_data['Uniswap'][plutocracy_report_data['Uniswap']['proposal_id'] == '0xfd3d3807bd2a6eda1327c311b83de235061d39ff1bdfb616c9f9b0d367c3ac2c'].iloc[0]['proposal_scores']

pd.DataFrame(
    {choice: [score, score_diff] for choice, score, score_diff in zip(eval(propsal_choices), eval(proposal_scores), proposal_score_diffrences)},
    index=["Scores", "Score Differeces"],
)

add Overnight Finance  add DeFiEdge     add none
Scores                     14335.930887  2.710600e+06  3556.089173
Score Differeces           10054.585076  2.708025e+06  3412.851497

One highly contentious proposal whose outcome did not change after filtering out whales was [this temperature check to choose which Eth <> BNB bridge to use for Uniswap v3 governance](https://snapshot.org/#/uniswap/proposal/0x6b8df360fdf73085b21fdf5eef9f85916fbde95621a3d454cb20fbe545ffc852). We see that even for the least popular choices, whales still contributed to the majority of the votes. We also observe that whales held the overwhelming majority of voting power for the runner up, "LayerZero", with 99.9% of votes being allocated to this choice. By comparison, "Wormhole" received 65.1% of voting power from whales.

In [9]:
propsal_choices = plutocracy_report_data['Uniswap'][plutocracy_report_data['Uniswap']['proposal_id'] == '0x6b8df360fdf73085b21fdf5eef9f85916fbde95621a3d454cb20fbe545ffc852'].iloc[0]['proposal_choices']
proposal_score_diffrences = score_diffrences_dfs["Uniswap"].loc["0x6b8df360fdf73085b21fdf5eef9f85916fbde95621a3d454cb20fbe545ffc852"]["score_diffrences"]
proposal_scores = plutocracy_report_data['Uniswap'][plutocracy_report_data['Uniswap']['proposal_id'] == '0x6b8df360fdf73085b21fdf5eef9f85916fbde95621a3d454cb20fbe545ffc852'].iloc[0]['proposal_scores']

pd.DataFrame(
    {choice: [score, score_diff] for choice, score, score_diff in zip(eval(propsal_choices), eval(proposal_scores), proposal_score_diffrences)},
    index=["Scores", "Score Differeces"],
)

Wormhole     LayerZero       Celer     deBridge
Scores            1.467367e+07  8.871274e+06  660.398099  2474.327951
Score Differeces  9.546794e+06  8.863809e+06  540.859826  2258.136409

In [10]:
score_diffrences_dfs["ENS"]

type  \
Proposal ID                                                         
0x41b3509b88e15677aa15680f48278517f794822fb9a79...  single-choice   
0xd7eff781be059513b5cd64d79e709abbbc653944c9a8c...  single-choice   
0x5788bf0f52ce82a1d3f7750a80f3001671ded49e4e023...  single-choice   
0xdaff050d30c77fe469da7eaa5b5bae0cb892f91a89c11...       approval   
0xa245dc7264ae072620434996e1c66d82c64cefc5a2edd...       approval   

                                                                                              choices  \
Proposal ID                                                                                             
0x41b3509b88e15677aa15680f48278517f794822fb9a79...                      ['For', 'Against', 'Abstain']   
0xd7eff781be059513b5cd64d79e709abbbc653944c9a8c...                      ['For', 'Against', 'Abstain']   
0x5788bf0f52ce82a1d3f7750a80f3001671ded49e4e023...                      ['For', 'Against', 'Abstain']   
0xdaff050d30c77fe469da7eaa5b5bae0cb892f91a89c11...  ['nick.eth', 'james.eth', 'accessor.eth', 'sim...   
0xa245dc7264ae072620434996e1c66d82c64cefc5a2edd...  ['thecap.eth (Cap)', 'slobo.eth', 'limes.eth',...   

                                                                                     score_diffrences  \
Proposal ID                                                                                             
0x41b3509b88e15677aa15680f48278517f794822fb9a79...       [1667210.963067633, 0.0, 137.12454090065714]   
0xd7eff781be059513b5cd64d79e709abbbc653944c9a8c...  [1666938.9315321625, 100.38682608595627, 141.9...   
0x5788bf0f52ce82a1d3f7750a80f3001671ded49e4e023...  [1446509.1013934365, 208348.53159207577, 13519...   
0xdaff050d30c77fe469da7eaa5b5bae0cb892f91a89c11...  [2319757.8152544936, 2319757.8152544936, 23197...   
0xa245dc7264ae072620434996e1c66d82c64cefc5a2edd...  [2278346.061608189, 2278346.0616081893, 227834...   

                                                             total_vp  \
Proposal ID                                                             
0x41b3509b88e15677aa15680f48278517f794822fb9a79...  1670185.073618348   
0xd7eff781be059513b5cd64d79e709abbbc653944c9a8c...  1670154.219044510   
0x5788bf0f52ce82a1d3f7750a80f3001671ded49e4e023...  1671687.088797375   
0xdaff050d30c77fe469da7eaa5b5bae0cb892f91a89c11...  2325021.309782755   
0xa245dc7264ae072620434996e1c66d82c64cefc5a2edd...  2282885.783012460   

                                                    has_changed_outcome  \
Proposal ID                                                               
0x41b3509b88e15677aa15680f48278517f794822fb9a79...                False   
0xd7eff781be059513b5cd64d79e709abbbc653944c9a8c...                False   
0x5788bf0f52ce82a1d3f7750a80f3001671ded49e4e023...                False   
0xdaff050d30c77fe469da7eaa5b5bae0cb892f91a89c11...                False   
0xa245dc7264ae072620434996e1c66d82c64cefc5a2edd...                False   

                                                   changed_outcome_winner  
Proposal ID                                                                
0x41b3509b88e15677aa15680f48278517f794822fb9a79...                    For  
0xd7eff781be059513b5cd64d79e709abbbc653944c9a8c...                    For  
0x5788bf0f52ce82a1d3f7750a80f3001671ded49e4e023...                    For  
0xdaff050d30c77fe469da7eaa5b5bae0cb892f91a89c11...               nick.eth  
0xa245dc7264ae072620434996e1c66d82c64cefc5a2edd...              slobo.eth

In [11]:
score_diffrences_dfs["Lido"]

type  \
Proposal ID                                                         
0x3436b98afacec300ec2c0f79bb8dfd16685ececeeb7a4...  single-choice   
0x32f6f095dc4f7b97665b560781d6e6795da11f9c82189...  single-choice   
0x7ac2431dc0eddcad4a02ba220a19f451ab6b064a0eaef...  single-choice   
0x7f19fed19b66df88b45d64c22f9905fbf716839a5052f...  single-choice   
0xcbf534335fe07c046caa933e1623ac38bfb3d1890ab82...  single-choice   

                                                                                              choices  \
Proposal ID                                                                                             
0x3436b98afacec300ec2c0f79bb8dfd16685ececeeb7a4...                                 ['For', 'Against']   
0x32f6f095dc4f7b97665b560781d6e6795da11f9c82189...                                 ['For', 'Against']   
0x7ac2431dc0eddcad4a02ba220a19f451ab6b064a0eaef...                                 ['For', 'Against']   
0x7f19fed19b66df88b45d64c22f9905fbf716839a5052f...                                     ['YAY', 'NAY']   
0xcbf534335fe07c046caa933e1623ac38bfb3d1890ab82...  ['1. For 6 new oracles, 6/11 set', '2. For 4 n...   

                                                                                     score_diffrences  \
Proposal ID                                                                                             
0x3436b98afacec300ec2c0f79bb8dfd16685ececeeb7a4...           [55124323.664074376, 2048432.8880345668]   
0x32f6f095dc4f7b97665b560781d6e6795da11f9c82189...             [57163748.21209293, 7563.113010643607]   
0x7ac2431dc0eddcad4a02ba220a19f451ab6b064a0eaef...             [57686947.46411716, 78.13038422652139]   
0x7f19fed19b66df88b45d64c22f9905fbf716839a5052f...             [54546195.85193226, 7525.505910950667]   
0xcbf534335fe07c046caa933e1623ac38bfb3d1890ab82...  [92934.92264371048, 52723122.2034864, 4535.772...   

                                                              total_vp  \
Proposal ID                                                              
0x3436b98afacec300ec2c0f79bb8dfd16685ececeeb7a4...  57176658.163439780   
0x32f6f095dc4f7b97665b560781d6e6795da11f9c82189...  57175091.606986538   
0x7ac2431dc0eddcad4a02ba220a19f451ab6b064a0eaef...  57690201.588526487   
0x7f19fed19b66df88b45d64c22f9905fbf716839a5052f...  54557517.318199627   
0xcbf534335fe07c046caa933e1623ac38bfb3d1890ab82...  52826368.429703690   

                                                    has_changed_outcome  \
Proposal ID                                                               
0x3436b98afacec300ec2c0f79bb8dfd16685ececeeb7a4...                False   
0x32f6f095dc4f7b97665b560781d6e6795da11f9c82189...                False   
0x7ac2431dc0eddcad4a02ba220a19f451ab6b064a0eaef...                False   
0x7f19fed19b66df88b45d64c22f9905fbf716839a5052f...                False   
0xcbf534335fe07c046caa933e1623ac38bfb3d1890ab82...                 True   

                                                            changed_outcome_winner  
Proposal ID                                                                         
0x3436b98afacec300ec2c0f79bb8dfd16685ececeeb7a4...                             For  
0x32f6f095dc4f7b97665b560781d6e6795da11f9c82189...                             For  
0x7ac2431dc0eddcad4a02ba220a19f451ab6b064a0eaef...                             For  
0x7f19fed19b66df88b45d64c22f9905fbf716839a5052f...                             YAY  
0xcbf534335fe07c046caa933e1623ac38bfb3d1890ab82...  1. For 6 new oracles, 6/11 set

In [12]:
score_diffrences_dfs["Frax"]

type  \
Proposal ID                                                    
0x1272233285b3b4ce026173ec4e2a59d6c8daba3d64b41...  weighted   
0x9cc997cf19f797919e40d87e17c7eed87aefa10279fbe...  weighted   
0x02aa360090c273c1218b84fe4b25e2be1d6a0210be141...  weighted   
0xe53269e838ab2ca0d8e09f8326d2a4ffb3425ecb1b39d...  weighted   
0xece8d5be8b180b54350c4bddee190e24e2849d233f8aa...  weighted   

                                                               choices  \
Proposal ID                                                              
0x1272233285b3b4ce026173ec4e2a59d6c8daba3d64b41...  ['For', 'Against']   
0x9cc997cf19f797919e40d87e17c7eed87aefa10279fbe...  ['For', 'Against']   
0x02aa360090c273c1218b84fe4b25e2be1d6a0210be141...  ['For', 'Against']   
0xe53269e838ab2ca0d8e09f8326d2a4ffb3425ecb1b39d...  ['For', 'Against']   
0xece8d5be8b180b54350c4bddee190e24e2849d233f8aa...  ['For', 'Against']   

                                                        score_diffrences  \
Proposal ID                                                                
0x1272233285b3b4ce026173ec4e2a59d6c8daba3d64b41...   [3805813.804935219]   
0x9cc997cf19f797919e40d87e17c7eed87aefa10279fbe...    [3805819.97255508]   
0x02aa360090c273c1218b84fe4b25e2be1d6a0210be141...   [3805825.454883847]   
0xe53269e838ab2ca0d8e09f8326d2a4ffb3425ecb1b39d...  [3805830.2519215182]   
0xece8d5be8b180b54350c4bddee190e24e2849d233f8aa...   [5792040.529852133]   

                                                             total_vp  \
Proposal ID                                                             
0x1272233285b3b4ce026173ec4e2a59d6c8daba3d64b41...  6086689.646356250   
0x9cc997cf19f797919e40d87e17c7eed87aefa10279fbe...  6073404.242473515   
0x02aa360090c273c1218b84fe4b25e2be1d6a0210be141...  6072772.098435019   
0xe53269e838ab2ca0d8e09f8326d2a4ffb3425ecb1b39d...  6072784.283546080   
0xece8d5be8b180b54350c4bddee190e24e2849d233f8aa...  9970590.463647339   

                                                    has_changed_outcome  \
Proposal ID                                                               
0x1272233285b3b4ce026173ec4e2a59d6c8daba3d64b41...                False   
0x9cc997cf19f797919e40d87e17c7eed87aefa10279fbe...                False   
0x02aa360090c273c1218b84fe4b25e2be1d6a0210be141...                False   
0xe53269e838ab2ca0d8e09f8326d2a4ffb3425ecb1b39d...                False   
0xece8d5be8b180b54350c4bddee190e24e2849d233f8aa...                False   

                                                   changed_outcome_winner  
Proposal ID                                                                
0x1272233285b3b4ce026173ec4e2a59d6c8daba3d64b41...                    For  
0x9cc997cf19f797919e40d87e17c7eed87aefa10279fbe...                    For  
0x02aa360090c273c1218b84fe4b25e2be1d6a0210be141...                    For  
0xe53269e838ab2ca0d8e09f8326d2a4ffb3425ecb1b39d...                    For  
0xece8d5be8b180b54350c4bddee190e24e2849d233f8aa...                    For

In [13]:
score_diffrences_dfs["Decentraland"]

type  \
Proposal ID                                                         
0xdfc9d87313c293cbde71adbed9d585ccd79d28ae1dde5...  single-choice   
0x7f6fed8c7645d1b793526564104e4f79864a9e30ae284...  single-choice   
0xbf2891a1e4490c0ba14e3897c239de06483a2758dff2c...  single-choice   
0x6576a4e2a22df3660d64a5af2a0d6c4819aa86ec40d42...  single-choice   
0xf574f75a5308ba8a35b5f6d2f3135dd185c5757bb52b5...  single-choice   

                                                                                         choices  \
Proposal ID                                                                                        
0xdfc9d87313c293cbde71adbed9d585ccd79d28ae1dde5...                                 ['yes', 'no']   
0x7f6fed8c7645d1b793526564104e4f79864a9e30ae284...     ['Yes', 'No', 'Invalid question/options']   
0xbf2891a1e4490c0ba14e3897c239de06483a2758dff2c...                                 ['yes', 'no']   
0x6576a4e2a22df3660d64a5af2a0d6c4819aa86ec40d42...  ['Unban', 'Ban', 'Invalid question/options']   
0xf574f75a5308ba8a35b5f6d2f3135dd185c5757bb52b5...                                 ['yes', 'no']   

                                                                             score_diffrences  \
Proposal ID                                                                                     
0xdfc9d87313c293cbde71adbed9d585ccd79d28ae1dde5...                                 [115.0, 0]   
0x7f6fed8c7645d1b793526564104e4f79864a9e30ae284...               [26021.000638214886, 0.0, 0]   
0xbf2891a1e4490c0ba14e3897c239de06483a2758dff2c...                   [888071.9501538787, 0.0]   
0x6576a4e2a22df3660d64a5af2a0d6c4819aa86ec40d42...  [809843.173033201, 0.0, 9364.50282589451]   
0xf574f75a5308ba8a35b5f6d2f3135dd185c5757bb52b5...     [569993.4788284205, 49981.66360630109]   

                                                            total_vp  \
Proposal ID                                                            
0xdfc9d87313c293cbde71adbed9d585ccd79d28ae1dde5...     142.356378308   
0x7f6fed8c7645d1b793526564104e4f79864a9e30ae284...   26949.339992801   
0xbf2891a1e4490c0ba14e3897c239de06483a2758dff2c...  904485.119040474   
0x6576a4e2a22df3660d64a5af2a0d6c4819aa86ec40d42...  844087.296455739   
0xf574f75a5308ba8a35b5f6d2f3135dd185c5757bb52b5...  631070.299142578   

                                                    has_changed_outcome  \
Proposal ID                                                               
0xdfc9d87313c293cbde71adbed9d585ccd79d28ae1dde5...                False   
0x7f6fed8c7645d1b793526564104e4f79864a9e30ae284...                False   
0xbf2891a1e4490c0ba14e3897c239de06483a2758dff2c...                False   
0x6576a4e2a22df3660d64a5af2a0d6c4819aa86ec40d42...                False   
0xf574f75a5308ba8a35b5f6d2f3135dd185c5757bb52b5...                False   

                                                   changed_outcome_winner  
Proposal ID                                                                
0xdfc9d87313c293cbde71adbed9d585ccd79d28ae1dde5...                    yes  
0x7f6fed8c7645d1b793526564104e4f79864a9e30ae284...                    Yes  
0xbf2891a1e4490c0ba14e3897c239de06483a2758dff2c...                    yes  
0x6576a4e2a22df3660d64a5af2a0d6c4819aa86ec40d42...                  Unban  
0xf574f75a5308ba8a35b5f6d2f3135dd185c5757bb52b5...                    yes

In [14]:
score_diffrences_dfs["Curve Finance"]

type  \
Proposal ID                                                         
0x0eb23ea0b877666ad3ddcd0d7da0114acdfe5ae6390b5...  single-choice   
0x43c689ef6b7b1772abc033cd87b7caa4159f0dc281f9a...  single-choice   
0xa11d4d1800b91746afe8ead922eb1d37225a32145adf3...  single-choice   
0x026a69e5c9780479332d88fa9779a3cec5a9479d05cc4...  single-choice   

                                                                                              choices  \
Proposal ID                                                                                             
0x0eb23ea0b877666ad3ddcd0d7da0114acdfe5ae6390b5...                                      ['Yes', 'No']   
0x43c689ef6b7b1772abc033cd87b7caa4159f0dc281f9a...  ['Do Nothing', '2Pool + current 3Pool', '2Pool...   
0xa11d4d1800b91746afe8ead922eb1d37225a32145adf3...                                      ['Yes', 'No']   
0x026a69e5c9780479332d88fa9779a3cec5a9479d05cc4...                                      ['yes', 'no']   

                                                                               score_diffrences  \
Proposal ID                                                                                       
0x0eb23ea0b877666ad3ddcd0d7da0114acdfe5ae6390b5...      [23462.994316173717, 511330.4724932205]   
0x43c689ef6b7b1772abc033cd87b7caa4159f0dc281f9a...  [0.0, 559204.7021513436, 2510541.702086272]   
0xa11d4d1800b91746afe8ead922eb1d37225a32145adf3...                       [26054297.07744241, 0]   
0x026a69e5c9780479332d88fa9779a3cec5a9479d05cc4...                      [1891.6266956075615, 0]   

                                                              total_vp  \
Proposal ID                                                              
0x0eb23ea0b877666ad3ddcd0d7da0114acdfe5ae6390b5...    574553.874246682   
0x43c689ef6b7b1772abc033cd87b7caa4159f0dc281f9a...   4272657.271177988   
0xa11d4d1800b91746afe8ead922eb1d37225a32145adf3...  26097304.209488451   
0x026a69e5c9780479332d88fa9779a3cec5a9479d05cc4...      6893.077810757   

                                                    has_changed_outcome  \
Proposal ID                                                               
0x0eb23ea0b877666ad3ddcd0d7da0114acdfe5ae6390b5...                 True   
0x43c689ef6b7b1772abc033cd87b7caa4159f0dc281f9a...                False   
0xa11d4d1800b91746afe8ead922eb1d37225a32145adf3...                False   
0x026a69e5c9780479332d88fa9779a3cec5a9479d05cc4...                False   

                                                           changed_outcome_winner  
Proposal ID                                                                        
0x0eb23ea0b877666ad3ddcd0d7da0114acdfe5ae6390b5...                            Yes  
0x43c689ef6b7b1772abc033cd87b7caa4159f0dc281f9a...  2Pool + current 3Pool + 4Pool  
0xa11d4d1800b91746afe8ead922eb1d37225a32145adf3...                            Yes  
0x026a69e5c9780479332d88fa9779a3cec5a9479d05cc4...                            yes

In [15]:
score_diffrences_dfs["Radicle"]

type  \
Proposal ID                                                         
0xf95b0e03c92f9f2df976cad6c5121383f4efc887162d0...  single-choice   
0x646e92794f8ea0abb333ed264dc8d8a3df7ceae909186...  single-choice   
bafkreia4tl25spzwtys4kfns3iiuj32n2wugyupa6adg6m...  single-choice   
0x16400b2c86d6ebdea246372182f18a17394435d9e31f9...          basic   
0x704025e577ff7e4e202985d342c30285c4d0d320a18c3...          basic   

                                                                                 choices  \
Proposal ID                                                                                
0xf95b0e03c92f9f2df976cad6c5121383f4efc887162d0...                    ['For', 'Against']   
0x646e92794f8ea0abb333ed264dc8d8a3df7ceae909186...              ['Yes', 'No', 'Abstain']   
bafkreia4tl25spzwtys4kfns3iiuj32n2wugyupa6adg6m...  ['For', 'Make No Change', 'Abstain']   
0x16400b2c86d6ebdea246372182f18a17394435d9e31f9...         ['For', 'Against', 'Abstain']   
0x704025e577ff7e4e202985d342c30285c4d0d320a18c3...         ['For', 'Against', 'Abstain']   

                                                              score_diffrences  \
Proposal ID                                                                      
0xf95b0e03c92f9f2df976cad6c5121383f4efc887162d0...              [2000000.0, 0]   
0x646e92794f8ea0abb333ed264dc8d8a3df7ceae909186...   [3730867.698436455, 0, 0]   
bafkreia4tl25spzwtys4kfns3iiuj32n2wugyupa6adg6m...   [2654186.698436455, 0, 0]   
0x16400b2c86d6ebdea246372182f18a17394435d9e31f9...  [3090365.0190418432, 0, 0]   
0x704025e577ff7e4e202985d342c30285c4d0d320a18c3...  [3090365.0190418432, 0, 0]   

                                                             total_vp  \
Proposal ID                                                             
0xf95b0e03c92f9f2df976cad6c5121383f4efc887162d0...  4133689.798082654   
0x646e92794f8ea0abb333ed264dc8d8a3df7ceae909186...  6163639.615623794   
bafkreia4tl25spzwtys4kfns3iiuj32n2wugyupa6adg6m...  6922974.673405429   
0x16400b2c86d6ebdea246372182f18a17394435d9e31f9...  4971478.525081047   
0x704025e577ff7e4e202985d342c30285c4d0d320a18c3...  4946306.104718991   

                                                    has_changed_outcome  \
Proposal ID                                                               
0xf95b0e03c92f9f2df976cad6c5121383f4efc887162d0...                False   
0x646e92794f8ea0abb333ed264dc8d8a3df7ceae909186...                False   
bafkreia4tl25spzwtys4kfns3iiuj32n2wugyupa6adg6m...                False   
0x16400b2c86d6ebdea246372182f18a17394435d9e31f9...                False   
0x704025e577ff7e4e202985d342c30285c4d0d320a18c3...                False   

                                                   changed_outcome_winner  
Proposal ID                                                                
0xf95b0e03c92f9f2df976cad6c5121383f4efc887162d0...                    For  
0x646e92794f8ea0abb333ed264dc8d8a3df7ceae909186...                    Yes  
bafkreia4tl25spzwtys4kfns3iiuj32n2wugyupa6adg6m...                    For  
0x16400b2c86d6ebdea246372182f18a17394435d9e31f9...                    For  
0x704025e577ff7e4e202985d342c30285c4d0d320a18c3...                    For

In [16]:
score_diffrences_dfs["Euler"]

type  \
Proposal ID                                                         
0x8721e87559f34b71b5ffc51cb9717d90677259d7a2b3d...  single-choice   
0xe16545bf8aa67399720dba057cf9d23f1e73caa8dad84...  single-choice   
0x3635964ebd6535eb25ea522b8aa5981942c2642491855...  single-choice   
0x4129f05639d2ca2f76cba392a625fa2ebe9b32964d29e...  single-choice   
0xeb9b8a930c53497494f0ba9af3b4411933b3aa2bbf845...  single-choice   

                                                                                              choices  \
Proposal ID                                                                                             
0x8721e87559f34b71b5ffc51cb9717d90677259d7a2b3d...  ['Delegate $SAFE to StableLab', 'Delegate $SAF...   
0xe16545bf8aa67399720dba057cf9d23f1e73caa8dad84...  ['YES - Implement changes', 'NO- Do not Implem...   
0x3635964ebd6535eb25ea522b8aa5981942c2642491855...  ['Yes - make rETH collateral', 'No - do not ma...   
0x4129f05639d2ca2f76cba392a625fa2ebe9b32964d29e...  ['YES - implement the changes ', 'NO -do not i...   
0xeb9b8a930c53497494f0ba9af3b4411933b3aa2bbf845...  ['YES - Implement the Changes ', 'NO - Do Noth...   

                                                                              score_diffrences  \
Proposal ID                                                                                      
0x8721e87559f34b71b5ffc51cb9717d90677259d7a2b3d...  [100000.0, 14950.714739293624, 0, 26366.0]   
0xe16545bf8aa67399720dba057cf9d23f1e73caa8dad84...                     [332920.55799487286, 0]   
0x3635964ebd6535eb25ea522b8aa5981942c2642491855...                   [332920.55799487286, 0.0]   
0x4129f05639d2ca2f76cba392a625fa2ebe9b32964d29e...                    [54944.92856057068, 0.0]   
0xeb9b8a930c53497494f0ba9af3b4411933b3aa2bbf845...    [47147.733648946596, 291603.84325557924]   

                                                            total_vp  \
Proposal ID                                                            
0x8721e87559f34b71b5ffc51cb9717d90677259d7a2b3d...  197217.723814994   
0xe16545bf8aa67399720dba057cf9d23f1e73caa8dad84...  388710.341413462   
0x3635964ebd6535eb25ea522b8aa5981942c2642491855...  397650.270054811   
0x4129f05639d2ca2f76cba392a625fa2ebe9b32964d29e...  105724.703030176   
0xeb9b8a930c53497494f0ba9af3b4411933b3aa2bbf845...  403627.500798613   

                                                    has_changed_outcome  \
Proposal ID                                                               
0x8721e87559f34b71b5ffc51cb9717d90677259d7a2b3d...                 True   
0xe16545bf8aa67399720dba057cf9d23f1e73caa8dad84...                False   
0x3635964ebd6535eb25ea522b8aa5981942c2642491855...                False   
0x4129f05639d2ca2f76cba392a625fa2ebe9b32964d29e...                False   
0xeb9b8a930c53497494f0ba9af3b4411933b3aa2bbf845...                 True   

                                                          changed_outcome_winner  
Proposal ID                                                                       
0x8721e87559f34b71b5ffc51cb9717d90677259d7a2b3d...                       Abstain  
0xe16545bf8aa67399720dba057cf9d23f1e73caa8dad84...       YES - Implement changes  
0x3635964ebd6535eb25ea522b8aa5981942c2642491855...    Yes - make rETH collateral  
0x4129f05639d2ca2f76cba392a625fa2ebe9b32964d29e...  YES - implement the changes   
0xeb9b8a930c53497494f0ba9af3b4411933b3aa2bbf845...  YES - Implement the Changes

In [17]:
score_diffrences_dfs["Hop"]

type  \
Proposal ID                                                         
0x29c0205077e5041a00b602c2e71a1b2ca5d84e2f7b4f0...  single-choice   
0xf1bf9327ad089d25e949b044d1673501acbeef96f1657...  single-choice   
0x6b5136469fd7b1478c0e10b95e30e8645a1cea8bb49ca...  single-choice   
0x77dfa74f2fa447137b13d8e5a6faaa7c9f2d1a23f86e8...  single-choice   
0xdf8dbd41184c7c401ee326c62ca6e253a8fa856b825b7...          basic   

                                                                          choices  \
Proposal ID                                                                         
0x29c0205077e5041a00b602c2e71a1b2ca5d84e2f7b4f0...       ['Yes', 'No', 'Abstain']   
0xf1bf9327ad089d25e949b044d1673501acbeef96f1657...       ['Yes', 'No', 'Abstain']   
0x6b5136469fd7b1478c0e10b95e30e8645a1cea8bb49ca...       ['Yes', 'No', 'Abstain']   
0x77dfa74f2fa447137b13d8e5a6faaa7c9f2d1a23f86e8...       ['Yes', 'No', 'Abstain']   
0xdf8dbd41184c7c401ee326c62ca6e253a8fa856b825b7...  ['For', 'Against', 'Abstain']   

                                                                                     score_diffrences  \
Proposal ID                                                                                             
0x29c0205077e5041a00b602c2e71a1b2ca5d84e2f7b4f0...                     [11802995.848545816, 0.0, 0.0]   
0xf1bf9327ad089d25e949b044d1673501acbeef96f1657...                     [13375900.370554667, 0.0, 0.0]   
0x6b5136469fd7b1478c0e10b95e30e8645a1cea8bb49ca...       [13366384.657650102, 6729.757727668588, 0.0]   
0x77dfa74f2fa447137b13d8e5a6faaa7c9f2d1a23f86e8...       [13371126.982221516, 6985.847465385696, 0.0]   
0xdf8dbd41184c7c401ee326c62ca6e253a8fa856b825b7...  [12960574.180947237, 10677.532957239824, 87717...   

                                                              total_vp  \
Proposal ID                                                              
0x29c0205077e5041a00b602c2e71a1b2ca5d84e2f7b4f0...  11985192.830100920   
0xf1bf9327ad089d25e949b044d1673501acbeef96f1657...  13563504.098747009   
0x6b5136469fd7b1478c0e10b95e30e8645a1cea8bb49ca...  13556220.919344630   
0x77dfa74f2fa447137b13d8e5a6faaa7c9f2d1a23f86e8...  13565274.220805701   
0xdf8dbd41184c7c401ee326c62ca6e253a8fa856b825b7...  13190658.527502781   

                                                    has_changed_outcome  \
Proposal ID                                                               
0x29c0205077e5041a00b602c2e71a1b2ca5d84e2f7b4f0...                False   
0xf1bf9327ad089d25e949b044d1673501acbeef96f1657...                False   
0x6b5136469fd7b1478c0e10b95e30e8645a1cea8bb49ca...                False   
0x77dfa74f2fa447137b13d8e5a6faaa7c9f2d1a23f86e8...                False   
0xdf8dbd41184c7c401ee326c62ca6e253a8fa856b825b7...                False   

                                                   changed_outcome_winner  
Proposal ID                                                                
0x29c0205077e5041a00b602c2e71a1b2ca5d84e2f7b4f0...                    Yes  
0xf1bf9327ad089d25e949b044d1673501acbeef96f1657...                    Yes  
0x6b5136469fd7b1478c0e10b95e30e8645a1cea8bb49ca...                    Yes  
0x77dfa74f2fa447137b13d8e5a6faaa7c9f2d1a23f86e8...                    Yes  
0xdf8dbd41184c7c401ee326c62ca6e253a8fa856b825b7...                    For